docs/proposal/20211207-delegated-storage.md

```diff
@@ -0,0 +1,554 @@
+- [Rejected ideas](#rejected-ideas)
+    + [Registering Storage Backends](#registering-storage-backends-1)
+    + [Workflow syntax](#workflow-syntax)
+    + [Backends implementation](#backends-implementation)
+- [Consequences](#consequences)
+
+<!-- tocstop -->
+
+## Motivation
+
+Capact stores the state in a form of TypeInstances with static data. That is problematic, as the data may quickly become outdated in case of any external change. For example, if you install Mattermost using Capact and delete Helm chart with Helm CLI, you will still be able to see Helm release TypeInstance of already not existing Helm release.
+
+Also, we should be able to provide a way to store sensitive data, such as credentials, securely. Currently, we store them as plaintext inside our database.
+
+The last reason, is that long-term we should replace Neo4j database for a more lightweight solution (like sqlite or PostgreSQL). Apart from being resource-hungry, Neo4j can be problematic also when it comes to licensing (GPL3). We believe that the pluggable back-ends concept could be a first step to abstract Neo4j and plug-in different storage backend.
+
+### Goal
+
+The main goal is to support the following use-cases:
+
+- Store and retrieve secrets using external, secure solutions
+    - Examples:
+        - user credentials (e.g. for PostgreSQL, GitLab)
+        - SSH key (e.g. for bastion)
+        - Kubeconfig (for external clusters)
+
+    - The following backends should be supported initially (sorted by priority):
+        1. Vault
+        1. Secrets encrypted with [SOPS](https://github.com/mozilla/sops) stored on Git repository
+        pluggable
+
+        Other backends:
+        - AWS Secrets Manager
+        - Google Key Management
+    
+- Store and retrieve dynamic data
+    - Examples:
+        - Kubernetes cluster (e.g. Flux's Helm releases or Kubernetes Secrets)
+        - Apps configuration (e.g. Mattermost config)
+        - External dependencies (e.g. S3 buckets)
+
+    - The following backends should be supported initially (sorted by priority):
+        1. Flux HelmRelease Custom Resources
+        1. Git repositories (e.g. GitLab projects)
+        1. S3 buckets (e.g. for external Terraform state)
+
+- Ability to manage such TypeInstances manually (via CLI and maybe UI) and as a part of Action
+    - Support such dynamic TypeInstance creation, update, and deletion
+    - Define GraphQL API and Implementation workflow syntax
+
+- Support for automatic TypeInstance creation and deletion
+    - React on events to create/delete such TypeInstances (e.g. Kubernetes events for any change for Flux's HelmReleases).
+
+- Support extensibility for upcoming backends
+
+Also, the additional, nice-to-have goals are:
+
+- Remove Neo4j dependency from Local Hub while preserving TypeInstance metadata, such as relations
+
+### Non-goal
+
+- Support external back-ends for Capact manifest storage (Public Hub)
+- Remove Neo4j dependency from Public Hub
+
+## Assumptions
```

Comment:
I'm not quite convinced to such idea, as IMO the Content Developer knows the best where a given value is stored, in case of dynamic TypeInstances. For static TypeInstances I agree, it's the Cluster Admin who decides where to put them.

So my assumption (which I also included) is: "Cluster Admin can configure default backend storage for static values.". For dynamic ones, it should be always Content Developer.

Reply:
That's good idea, I will address it in the proposal 👍